dig.go

```diff
@@ -648,12 +696,14 @@ func shallowCheckDependencies(c containerStore, p param) error {
 
 		if ns := c.getValueProviders(ps.Name, ps.Type); len(ns) == 0 && !ps.Optional {
 			missing = append(missing, newErrMissingType(c, key{name: ps.Name, t: ps.Type}))
+			missingNodes = append(missingNodes, ps.DotParam()...)
 		}
 
 		return true
 	}))
 
 	if len(missing) > 0 {
+		c.failNodes(missingNodes, 0)
```

Comment:
What does it mean when id is 0? Use of a magic value like this suggests we
should do something else here. Maybe `c.failMissingNodes(missingNodes)`?

Reply:
I've added a `missingNodes` function that deals with nodes with no constructors.